### PR TITLE
Indent body and type when printing local definitions over several lines

### PR DESCRIPTION
**Pull request by herbelin**
Assume:
```coq
Theorem T (long_long_long_long_long_long_long_long_ident := fun x : nat => x) : True.
```
Before the PR, it is printed
```coq
  long_long_long_long_long_long_long_long_ident :=
  fun x : nat => x : nat -> nat
  ============================
  True
```
After, it is printed:
```coq
  long_long_long_long_long_long_long_long_ident :=
    fun x : nat => x : nat -> nat
  ============================
  True
```
The PR also applies this indentation to other similar places.

#### Overlays (to be merged before the current PR)

* [x] https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/549

### CODE REVIEW

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
@herbelin apparently this requires modifying a few output tests

**Comment by herbelin** [comment]
> apparently this requires modifying a few output tests

OK. I seized then the opportunity to also "improve" the printing of multiple names of same type.

File: test-suite/output/Cases.out [hunk 1]
```diff
@@ -102,11 +102,10 @@ Arguments lem3 k
...
            end : nat
-  e,
-  e0 := match x + 0 as y return (y = y) with
-        | 0 => eq_refl
-        | S n => eq_refl
-        end : x + 0 = x + 0
+  e, e0 := match x + 0 as y return (y = y) with
+           | 0 => eq_refl
+           | S n => eq_refl
+           end : x + 0 = x + 0
```
Comment (SkySkimmer) [inline test-suite/output/Cases.out:108]: Could we get the `match` on a new line so that the branches don't get that indented? eg
```suggestion
  e, e0 :=
    match x + 0 as y return (y = y) with
    | 0 => eq_refl
    | S n => eq_refl
    end : x + 0 = x + 0
```

File: test-suite/output/PrintAssumptions.out [hunk 2]
```diff
@@ -10,16 +10,16 @@ Axioms:
 seq relies on definitional UIP.
 Axioms:
 extensionality
-  : forall (P Q : Type) (f g : P -> Q), (forall x : P, f x = g x) -> f = g
+   : forall (P Q : Type) (f g : P -> Q), (forall x : P, f x = g x) -> f = g
```
Comment (proux01) [inline test-suite/output/PrintAssumptions.out:13]: Is this change intended? No big deal, but I feel like it was fine before.

File: test-suite/output/Cases.out (hunk 1 above)
Comment (herbelin) [reply test-suite/output/Cases.out:108]: One would need to give a priority on when to break between different nested boxes and I'm unsure to know how to control that :(

File: test-suite/output/PrintAssumptions.out (hunk 2 above)
Comment (herbelin) [reply test-suite/output/PrintAssumptions.out:13]: It is unrelated to the other changes but it was intended in the commit. This was the only case indenting with 1 space while other indentations are using 2 spaces.

But I agree that 1 is esthetically fine also, so I can revert if you prefer.

Comment (herbelin) [reply test-suite/output/PrintAssumptions.out:13]: Actually, you're right, I made a mistake. There is already one space before ":", so no need for one more. Will change when full CI is finished.

File: test-suite/output/Cases.out (hunk 1 above)
Comment (proux01) [reply test-suite/output/Cases.out:108]: I would have expected an enclosing box to be enough:, like
"@[<hov 2> @[<hov 0>comma separated variables@] :=@ body@]"
(OCaml format syntax).

Comment (herbelin) [reply test-suite/output/Cases.out:108]: This is already what we have I think (in `Pp` syntax):
```ocaml
  let pids = hov 0 (prlist_with_sep pr_comma (fun id -> pr_id id.binder_name) ids) in
  let pt = pr_ltype_env env sigma typ in
  let ptyp = (str" : " ++ pt) in
  hov 2 (pids ++ pbody ++ ptyp)
```

Comment (proux01) [reply test-suite/output/Cases.out:108]: But in pbody I read `str" := " ++ pb` whereas I would expect `str" :=" ++ spc () ++ pb`

Comment (herbelin) [reply test-suite/output/Cases.out:108]: Sorry, I missed the break after `:=` in your code. It is not in the Coq code and adding it does what @SkySkimmer is asking.

File: printing/printer.ml [hunk 3]
```diff
@@ -610,7 +610,7 @@ let pr_evgl_sign env sigma (evi : undefined evar_info) =
 let pr_evar sigma (evk, evi) =
   let env = Global.env () in
   let pegl = pr_evgl_sign env sigma evi in
-  hov 0 (pr_existential_key env sigma evk ++ str " : " ++ pegl)
+  hov 2 (pr_existential_key env sigma evk ++ str " : " ++ pegl)
```
Comment (proux01) [inline printing/printer.ml:613]: I guess it should be
```suggestion
  hov 2 (pr_existential_key env sigma evk ++ spc () ++ str ": " ++ pegl)
```

File: printing/printer.ml [hunk 4]
```diff
@@ -344,12 +344,12 @@ let pr_compacted_decl env sigma decl =
        (* Force evaluation *)
        let pb = pr_lconstr_env ~inctx:true env sigma c in
        let pb = if isCast c then surround pb else pb in
-       ids, (str" := " ++ pb ++ cut ()), typ
+       ids, (str" :=" ++ spc () ++ pb ++ cut ()), typ
   in
-  let pids = prlist_with_sep pr_comma (fun id -> pr_id id.binder_name) ids in
+  let pids = hov 0 (prlist_with_sep pr_comma (fun id -> pr_id id.binder_name) ids) in
   let pt = pr_ltype_env env sigma typ in
   let ptyp = (str" : " ++ pt) in
```
Comment (proux01) [inline printing/printer.ml:351]: I guess this should be
```suggestion
  let ptyp = (spc () ++ str": " ++ pt) in
```

File: printing/printer.ml (hunk 3 above)
Comment (proux01) [reply printing/printer.ml]: I pushed a fix

File: printing/printer.ml (hunk 4 above)
Comment (herbelin) [reply printing/printer.ml]: You prefer for long lines when the `:` is on the next line (as in the new tests)? If ok for you, that's ok to me too.

Comment (proux01) [reply printing/printer.ml]: On the testcases, it seems debatable because the identifiers are very short, but would they be longer, we'd end up with
```
long_long_identifier : thing indentend
                                       with crazy margin
                                       in the middle
                                       of your screen
```
so yes, I think it's better overall

Comment (herbelin) [reply printing/printer.ml]: An alternative is to cut after the `:` (like in `pr_rel_decl`) so that there is a choice between:
```
long_long_identifier
  : thing indented
                   with crazy margin
                   in the middle
                   of your screen
```
and
```
long_long_identifier :
  thing indented
                 with crazy margin
                 in the middle
                 of your screen
```

Comment (proux01) [reply printing/printer.ml]: Ok, done that

Comment (herbelin) [reply printing/printer.ml]: All output looks nice to me.

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@proux01 are you going to self assign?


Looks like the iris output tests are sensitive to this change, cc @RalfJung, @robbertkrebbers

**Comment by RalfJung** [comment]
The Iris issue should be fixed now, sorry for the inconvenience.

**Comment by proux01** [comment]
Thanks a lot

**Comment by proux01** [comment]
@coqbot merge now
